src/utils/body-limit.ts

```diff
@@ -0,0 +1,86 @@
+import type { H3Event } from "../event.ts";
+import type { H3Plugin } from "../types/h3.ts";
+import { HTTPError } from "../error.ts";
+import { defineMiddleware } from "../middleware.ts";
+
+export interface BodySizeLimitOptions {
+  /**
+   * Maximum allowed body size in bytes
+   */
+  maxSize: number;
+  /**
+   * Routes to apply the limit to (optional)
+   * If not specified, applies to all routes
+   */
+  routes?: Array<string | RegExp>;
+  /**
+   * Routes to exclude from the limit (optional)
+   */
+  exclude?: Array<string | RegExp>;
+}
+
+/**
+ * Define a plugin that limits request body size
+ *
+ * @example
+ * ```js
+ * import { defineBodySizeLimitPlugin } from "h3";
+ *
+ * const bodySizeLimit = defineBodySizeLimitPlugin({
+ *   maxSize: 1024 * 1024, // 1MB
+ *   routes: ["/api/upload", /^\/api\/files/],
+ *   exclude: ["/api/large-upload"]
+ * });
+ *
+ * app.register(bodySizeLimit);
+ * ```
+ */
+export function defineBodySizeLimitPlugin(
```

Comment:
It should be a `middleware` not a plugin to apply in places needed.

With middleware, pattern and route matching can be defined also on definition therefore we do not need to (re)implement logic inside this utility.